### PR TITLE
Add Unmanaged FSx multiple AZ validator.

### PR DESCRIPTION
**Pull request by NSsirena**
Unmanaged FSx can be associated to AZs that differ from the ones configured in the queues. In such cases we inform the customers that they may incur in increased latency and costs.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

### Description of changes
* Add few properties in BaseSharedFsx to 
    * retrieve the list of subnets
    * know if the FS is unmanaged
* Add Unmanaged FSx multiple AZ validator that
    * compares the set of subnets of the queue and the subnets of the FS
    * informs of possible increased latency and costs in case of mismatch

### Tests
* Unit tests.

### References
None

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] ~~Check if documentation is impacted by this change.~~

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/config/cluster_config.py [hunk 1]
```diff
@@ -371,17 +372,27 @@ def __init__(self, mount_dir: str, name: str):
...
-        if not self.__file_system_data and self.file_system_id:
+        if not self.__file_system_data and self.unmanaged:
             self.__file_system_data = AWSApi.instance().fsx.get_file_systems_info([self.file_system_id])[0]
         return self.__file_system_data
 
+    @property
+    def file_system_subnets(self):
+        """Return list of subnets associated to FSx."""
+        return self.file_system_data.subnet_ids if self.unmanaged else []
+
```
Comment (jdeamicis) [inline cli/src/pcluster/config/cluster_config.py]: This implementation/name might cause confusion in the future: here you are returning the subnet IDs only if the FSx is unmanaged. Is it wise to create this difference?

Comment (jdeamicis) [reply cli/src/pcluster/config/cluster_config.py]: I see you are probably using a pre-existing pattern here. Feel free to dismiss this comment.

Comment (jdeamicis) [reply cli/src/pcluster/config/cluster_config.py]: Also, is a new property absolutely necessary, since this information already exists in `file_system_data`?

File: cli/src/pcluster/validators/cluster_validators.py [hunk 2]
```diff
@@ -701,6 +701,25 @@ def _validate(self, queues, new_storage_count):
...
+    """
+    Unmanaged FSx Storage Vs Multiple Subnets validator.
+
+    Unmanaged FSx volumes can exist in AZ that are different from the ones defined in queues configuration.
+    In these cases we notify customers that they may incur in increased latency and costs.
+    """
+
+    def _validate(self, queues, fsx_subnets):
+        for queue in queues:
+            if not set(queue.networking.subnet_ids) == set(fsx_subnets):
```
Comment (jdeamicis) [inline cli/src/pcluster/validators/cluster_validators.py]: Are we sure ONTAP is multi-zonal in the sense that it supports different mount targets on different AZs? If not, would this be the right logic for a multizonal ONTAP?

File: cli/tests/pcluster/validators/test_cluster_validators.py [hunk 3]
```diff
@@ -1724,6 +1725,176 @@ def test_new_storage_multiple_subnets_validator(queues, new_storage_count, failu
...
+                        subnet_ids=["subnet-1", "subnet-2"],
+                    ),
+                ),
+            ],
+            ["subnet-1", "subnet-2"],
+            None,
+            "",
+        ),
+    ],
+)
```
Comment (jdeamicis) [inline cli/tests/pcluster/validators/test_cluster_validators.py:1976]: Can you please make one of these cases use the `SchedulerPluginQueueNetworking` class?

**Review by jdeamicis — COMMENTED** [review]
Left some comments to discuss about.

File: cli/src/pcluster/config/cluster_config.py (hunk 1 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py]: I'd prefer not to ..however `file_system_data` is retrieved only if the FS is unmanaged. This is why I put the check here. 
Moreover I created a property to make it explicit since the previous check wasn't so clear.

I thought about generalizing this also for the managed case, but this was out-of-scope for this PR and also I didn't want to break existing behaviors.

File: cli/src/pcluster/config/cluster_config.py [hunk 4]
```diff
@@ -371,17 +372,27 @@ def __init__(self, mount_dir: str, name: str):
     def _register_validators(self, context: ValidatorContext = None):  # noqa: D102 #pylint: disable=unused-argument
         self._register_validator(SharedStorageNameValidator, name=self.name)
 
+    @property
+    def unmanaged(self):
```
Comment (lukeseawalker) [inline cli/src/pcluster/config/cluster_config.py]: TIP - tipically we use "is_" prefix to describe that a boolean value is returned.

Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py]: Got it.
I'll update the property name.

File: cli/tests/pcluster/validators/test_cluster_validators.py (hunk 3 above)
Comment (NSsirena) [reply cli/tests/pcluster/validators/test_cluster_validators.py:1976]: Sure, thanks for catching this.

File: cli/src/pcluster/validators/cluster_validators.py (hunk 2 above)
Comment (jdeamicis) [reply cli/src/pcluster/validators/cluster_validators.py]: Additional comment (more important): this poses the "constraint" (although it's just an INFO) to use the same subnets between the queue and the FSx. We should check the AZs, not the subnets.

File: cli/src/pcluster/config/cluster_config.py (hunk 1 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py]: Discussed this with Jacopo and agreed to proceed with this approach since it's extensively used.

File: cli/src/pcluster/validators/cluster_validators.py (hunk 2 above)
Comment (NSsirena) [reply cli/src/pcluster/validators/cluster_validators.py]: I totally agree.
I fixed the check and now uses AZ list instead of subnet list.

File: cli/tests/pcluster/validators/test_cluster_validators.py (hunk 3 above)
Comment (NSsirena) [reply cli/tests/pcluster/validators/test_cluster_validators.py:1976]: Fixed

File: cli/src/pcluster/config/cluster_config.py (hunk 4 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py]: Fixed

File: cli/src/pcluster/config/cluster_config.py [hunk 5]
```diff
@@ -371,17 +372,38 @@ def __init__(self, mount_dir: str, name: str):
...
+        """Return list of subnets associated to existing FSx."""
+        return self.file_system_data.subnet_ids if self.is_unmanaged else []
+
+    @property
+    def file_system_azs(self):
+        """Return list of AZ associated to existing FSx."""
+        azs = []
+        if self.is_unmanaged:
+            mapping = AWSApi.instance().ec2.get_subnets_az_mapping(self.file_system_subnets)
+            for _subnet_id, availability_zone in mapping.items():
```
Comment (lukeseawalker) [inline cli/src/pcluster/config/cluster_config.py]: if `_subnet_id` is not used, you can use
```
for _, availability_zone in mapping.items():
```

File: cli/src/pcluster/config/cluster_config.py [hunk 6]
```diff
@@ -371,17 +372,38 @@ def __init__(self, mount_dir: str, name: str):
...
         return self.__file_system_data
 
+    @property
+    def file_system_subnets(self):
+        """Return list of subnets associated to existing FSx."""
+        return self.file_system_data.subnet_ids if self.is_unmanaged else []
+
+    @property
+    def file_system_azs(self):
+        """Return list of AZ associated to existing FSx."""
```
Comment (lukeseawalker) [inline cli/src/pcluster/config/cluster_config.py]: since there is an az_list method, what about calling this `file_system_az_list` for coherence?

File: cli/src/pcluster/config/cluster_config.py [hunk 7]
```diff
@@ -1379,6 +1405,13 @@ def _validate_new_storage_multiple_subnets(self, compute_subnet_ids, new_storage
             compute_subnet_ids=compute_subnet_ids,
             new_storage_count=new_storage_count,
         )
+        for storage in self.shared_storage:
+            if isinstance(storage, (SharedFsxLustre, ExistingFsxOpenZfs, ExistingFsxOntap)) and storage.is_unmanaged:
+                self._register_validator(
+                    UnmanagedFsxMultiAzValidator,
+                    queues=queues,
+                    fsx_subnets=storage.file_system_subnets,
```
Comment (lukeseawalker) [inline cli/src/pcluster/config/cluster_config.py]: isn't `UnmanagedFsxMultiAzValidator` expecting a list of AZ instead of a list of subnets?

File: cli/src/pcluster/validators/cluster_validators.py [hunk 8]
```diff
@@ -701,6 +701,25 @@ def _validate(self, compute_subnet_ids, new_storage_count):
...
+    """
+    Unmanaged FSx Storage Vs Multiple Subnets validator.
+
+    Unmanaged FSx volumes can exist in AZ that are different from the ones defined in queues configuration.
+    In these cases we notify customers that they may incur in increased latency and costs.
+    """
+
+    def _validate(self, queues, fsx_az_list):
+        for queue in queues:
+            if not set(queue.networking.az_list) == set(fsx_az_list):
```
Comment (lukeseawalker) [inline cli/src/pcluster/validators/cluster_validators.py]: In principle, they way az_list list are built, there should not be duplicated elements, right?

File: cli/tests/pcluster/validators/test_cluster_validators.py [hunk 9]
```diff
@@ -1790,6 +1792,153 @@ def test_new_storage_multiple_subnets_validator(queues, new_storage_count, failu
...
+                ),
+            ],
+            {"subnet-1": "us-east-1a", "subnet-2": "us-east-1b"},
+            ["us-east-1b"],
+            FailureLevel.INFO,
+            "Your configuration for Queue multi-az-queue-partial-match includes multiple subnets and external "
+            "shared storage configuration. Accessing a shared storage from different AZs can lead to increased "
+            "latency and costs.",
+        ),
+        (
```
Comment (lukeseawalker) [inline cli/tests/pcluster/validators/test_cluster_validators.py:1912]: duplicated test

File: cli/tests/pcluster/validators/test_cluster_validators.py [hunk 10]
```diff
@@ -1790,6 +1792,153 @@ def test_new_storage_multiple_subnets_validator(queues, new_storage_count, failu
...
+                ),
+            ],
+            {"subnet-1": "us-east-1a", "subnet-2": "us-east-1b"},
+            ["us-east-1a", "us-east-1b"],
+            None,
+            "",
+        ),
+    ],
+)
+def test_unmanaged_fsx_multiple_az_validator(
```
Comment (lukeseawalker) [inline cli/tests/pcluster/validators/test_cluster_validators.py:1977]: can you add a test with len(SlurmQueue) > 1?

File: cli/src/pcluster/config/cluster_config.py [hunk 11]
```diff
@@ -1379,6 +1405,13 @@ def _validate_new_storage_multiple_subnets(self, compute_subnet_ids, new_storage
             compute_subnet_ids=compute_subnet_ids,
             new_storage_count=new_storage_count,
         )
+        for storage in self.shared_storage:
+            if isinstance(storage, (SharedFsxLustre, ExistingFsxOpenZfs, ExistingFsxOntap)) and storage.is_unmanaged:
+                self._register_validator(
+                    UnmanagedFsxMultiAzValidator,
+                    queues=queues,
```
Comment (lukeseawalker) [inline cli/src/pcluster/config/cluster_config.py:1414]: `queues` is not defined

File: cli/src/pcluster/config/cluster_config.py [hunk 12]
```diff
@@ -371,17 +372,38 @@ def __init__(self, mount_dir: str, name: str):
...
+        """Return list of subnets associated to existing FSx."""
+        return self.file_system_data.subnet_ids if self.is_unmanaged else []
+
+    @property
+    def file_system_availability_zones(self):
+        """Return list of AZ associated to existing FSx."""
+        availability_zones = []
+        if self.is_unmanaged:
+            mapping = AWSApi.instance().ec2.get_subnets_az_mapping(self.file_system_subnets)
+            for _, availability_zone in mapping.items():
```
Comment (lukeseawalker) [inline cli/src/pcluster/config/cluster_config.py]:
```
for availability_zone in mapping.values():
```

File: cli/src/pcluster/config/cluster_config.py [hunk 13]
```diff
@@ -1373,12 +1401,19 @@ def _validate_mount_dirs(self):
...
             ManagedFsxMultiAzValidator,
             compute_subnet_ids=compute_subnet_ids,
             new_storage_count=new_storage_count,
         )
+        for storage in self.shared_storage:
+            if isinstance(storage, (SharedFsxLustre, ExistingFsxOpenZfs, ExistingFsxOntap)) and storage.is_unmanaged:
+                self._register_validator(
+                    UnmanagedFsxMultiAzValidator,
+                    queues=queues,
+                    fsx_subnets=storage.file_system_availability_zones,
```
Comment (lukeseawalker) [inline cli/src/pcluster/config/cluster_config.py]: wrong param name, subnet != az

File: cli/src/pcluster/config/cluster_config.py (hunk 12 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py]: 👍 ..I'll fix it

File: cli/src/pcluster/config/cluster_config.py (hunk 13 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py]: oops! ...good catch!
wonder why it was working...

File: cli/src/pcluster/validators/cluster_validators.py [hunk 14]
```diff
@@ -701,6 +701,25 @@ def _validate(self, compute_subnet_ids, new_storage_count):
...
+                self._add_failure(
+                    "Your configuration for Queue {0} includes multiple subnets and external shared storage "
```
Comment (lukeseawalker) [inline cli/src/pcluster/validators/cluster_validators.py]: quote the values, i.e. `'{0}'`

File: cli/tests/pcluster/validators/test_cluster_validators.py [hunk 15]
```diff
@@ -1790,6 +1792,167 @@ def test_new_storage_multiple_subnets_validator(queues, new_storage_count, failu
...
+                    compute_resources=[],
+                    networking=SlurmQueueNetworking(
+                        subnet_ids=["subnet-1"],
+                    ),
+                ),
+            ],
+            [{"subnet-2": "us-east-1b"}, {"subnet-1": "us-east-1a"}],
+            ["us-east-1a"],
+            FailureLevel.INFO,
+            "Your configuration for Queue single-az-different-subnet-queue includes multiple subnets and external "
```
Comment (lukeseawalker) [inline cli/tests/pcluster/validators/test_cluster_validators.py]: add quoting to the value, i.e. `'single-az-different-subnet-queue'`

File: cli/tests/pcluster/validators/test_cluster_validators.py [hunk 16]
```diff
@@ -1790,6 +1792,167 @@ def test_new_storage_multiple_subnets_validator(queues, new_storage_count, failu
...
+                    networking=SchedulerPluginQueueNetworking(
+                        subnet_ids=["subnet-1", "subnet-2"],
+                    ),
+                ),
+            ],
+            [{"subnet-1": "us-east-1a", "subnet-2": "us-east-1b"}],
+            ["us-east-1a", "us-east-1b"],
+            None,
+            "",
+        ),
```
Comment (lukeseawalker) [inline cli/tests/pcluster/validators/test_cluster_validators.py:1974]: missing test case with more than 1 queue where none of the queue is in the same AZ of the FS. This to trigger the path of multiple failure messages

File: cli/src/pcluster/validators/cluster_validators.py (hunk 14 above)
Comment (NSsirena) [reply cli/src/pcluster/validators/cluster_validators.py]: I had quotes. I removed them since in the messages reported in the doc the values were unquoted.

Anyway I'll bring back the quotes.

File: cli/tests/pcluster/validators/test_cluster_validators.py (hunk 15 above)
Comment (NSsirena) [reply cli/tests/pcluster/validators/test_cluster_validators.py]: Same as above.

File: cli/tests/pcluster/validators/test_cluster_validators.py [hunk 17]
```diff
@@ -1790,6 +1792,167 @@ def test_new_storage_multiple_subnets_validator(queues, new_storage_count, failu
...
+                    compute_resources=[],
+                    networking=SlurmQueueNetworking(
+                        subnet_ids=["subnet-1", "subnet-2"],
+                    ),
+                ),
+            ],
+            [{"subnet-1": "us-east-1a"}, {"subnet-1": "us-east-1a", "subnet-2": "us-east-1b"}],
+            ["us-east-1a", "us-east-1b"],
+            FailureLevel.INFO,
+            "Your configuration for Queue 'one-az-match-queue' includes multiple subnets and external "
```
Comment (lukeseawalker) [inline cli/tests/pcluster/validators/test_cluster_validators.py]: there should be no error in this case.
FS is in AZ ["us-east-1a", "us-east-1b"], so cluster will be able to mount the FS in the same AZ

Comment (NSsirena) [reply cli/tests/pcluster/validators/test_cluster_validators.py]: This is true IF and OLNY IF the data in each AZ is an exact replica and so a compute node in `us-east-1a` will connect to the FS mount point in `us-east-1a` and all the data will be local.

If, somehow, this is not true data will be transferred from one AZ to another and this will have an impact on the latency while accessing the data and possibly cost.

Comment (NSsirena) [reply cli/tests/pcluster/validators/test_cluster_validators.py]: We discussed this topic in a call a decided to slightly change the logic of the validator.

**Review by lukeseawalker — APPROVED** [review]
